shared/src/main/scala/mlscript/JSBackend.scala

```diff
@@ -517,13 +484,12 @@ class JSBackend(allowUnresolvedSymbols: Boolean) {
     JSClassDecl(classSymbol.runtimeName, fields, base, members, traits)
   }
 
-  protected def translateSelfDeclaration(self: ValueSymbol): Ls[JSStmt] =
-    if (visitedSymbols(self)) {
-      visitedSymbols -= self
-      JSConstDecl(self.runtimeName, JSIdent("this")) :: Nil
+  protected def translateQualifierDeclaration(qualifier: ValueSymbol): Ls[JSStmt] =
+    if (visitedSymbols(qualifier)) {
+      visitedSymbols -= qualifier
```

Comment:
What's going on here? Please justify this line with a comment or use a cleaner approach. Think about what would happen if the symbol was truly visited first, and then later removed by this line.

Reply:
For each scope, we create a qualifier symbol (i.e., an alias for `this`). We generate the actual `const qualifier = this;` only if the symbol is used in the body. This method would be called at the end of the member generation, so removing or not does not effect the result. Since this method can also be called somewhere else and it may lead to unexpected results, this line has been removed.

Reply:
In fact, why is `visitedSymbols` a global variable? It has no business being reused all over the place and storing every symbol... that's just smelly design. At this point, why not just store a Boolean flag in the symbol itself? It's equally smelly, but at least it will be much more efficient.

Reply:
Actually, here I just reused the existing code: [shared/src/main/scala/mlscript/JSBackend.scala](https://github.com/hkust-taco/mlscript/blob/mlscript/shared/src/main/scala/mlscript/JSBackend.scala#L22).

Reply:
Sure. Still, while we're at it, could you plz make this a mutable flag in the symbol instead?

Reply:
In general the principle is that it's better to leave the code in a better state than we found it in ;)

Reply:
@NeilKleistGao Are we going to remove this `visitedSymbols`? It would make sense, since you removed the other one.

Reply:
I think this `visitedSymbols` has been removed 🤔

Reply:
Oh yeah sorry I looked at the wrong diff!